src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,312 @@
+                    await UpdateReferenceAsync(project, reference, cancellationToken).ConfigureAwait(false);
                 }
             }
 
-            private void ReportCount()
+            private async Task UpdateReferenceAsync(
+                Project project, PortableExecutableReference reference, CancellationToken cancellationToken)
             {
-                var sourceSymbolCount = _symbolCountByProjectMap.Sum(kv => kv.Value);
-                Logger.Log(FunctionId.Run_Environment, KeyValueLogMessage.Create(m => m["SourceSymbolCount"] = sourceSymbolCount));
+                var key = GetReferenceKey(reference);
+                if (key == null)
+                {
+                    return;
+                }
 
-                // we only report it once
-                _symbolCountByProjectMap = null;
-                _solutionId = null;
-            }
+                DateTime lastWriteTime;
+                if (!TryGetLastWriteTime(key, out lastWriteTime))
+                {
+                    // Couldn't get the write time.  Just ignore this reference.
+                    return;
+                }
 
-            private void RecordCount(ProjectId id, int count)
-            {
-                if (_symbolCountByProjectMap == null)
+                MetadataInfo metadataInfo;
+                if (_metadataPathToInfo.TryGetValue(key, out metadataInfo) && metadataInfo.TimeStamp == lastWriteTime)
                 {
+                    // We've already computed and cached the info for this reference.
                     return;
                 }
 
-                _symbolCountByProjectMap[id] = count;
+                var compilationService = project.LanguageServices.GetService<ICompilationFactoryService>();
+                var compilation = compilationService.CreateCompilation("TempAssembly", compilationService.GetDefaultCompilationOptions())
+                    .WithReferences(reference);
+
+                var assembly = compilation.GetAssemblyOrModuleSymbol(reference) as IAssemblySymbol;
+                if (assembly == null)
+                {
+                    return;
+                }
+
+                var info = await SymbolTreeInfo.TryGetInfoForMetadataAssemblyAsync(project.Solution, assembly, reference, loadOnly: false, cancellationToken: cancellationToken).ConfigureAwait(false);
+                metadataInfo = metadataInfo ?? new MetadataInfo(lastWriteTime, info, new HashSet<ProjectId>());
+
+                // Keep track that this dll is referenced by this project.
+                metadataInfo.ReferencingProjects.Add(project.Id);
+
+                _metadataPathToInfo.AddOrUpdate(key, metadataInfo, (_1, _2) => metadataInfo);
+            }
+
+            public override void RemoveProject(ProjectId projectId)
+            {
+                ProjectInfo info;
+                _projectToInfo.TryRemove(projectId, out info);
+                _idTolastSeenProject.Remove(projectId);
```

Comment:
I remain concerned about keeping fully fledged `Project` objects around indefinitely, but maybe they are lightweight enough if as long as nothing is strongly keeping their `Compilation`s alive?  @mattwar, @heejaechang, @KevinH-MS Any thoughts on the steady state memory usage here.